app/vmselect/netstorage/netstorage.go

```diff
@@ -3286,6 +3270,29 @@ func (pnc *perNodeCounter) GetTotal() uint64 {
 // See https://github.com/golang/go/blob/704401ffa06c60e059c9e6e4048045b4ff42530a/src/runtime/malloc.go#L11
 const maxFastAllocBlockSize = 32 * 1024
 
+// execSearchQueryRequest executes processSearchQuery for each searchQuery tenant.
+func execSearchQueryRequest(qt *querytracer.Tracer, sq *storage.SearchQuery, workerID uint, sn *storageNode, f func(mb *storage.MetricBlock, workerID uint) error, deadline searchutils.Deadline) error {
+	var requestData []byte
+
+	for i := range sq.TenantTokens {
+		requestData = sq.TenantTokens[i].Marshal(requestData)
+		requestData = sq.MarshalWithoutTenant(requestData)
+		qtL := qt
+		if sq.IsMultiTenant && qt.Enabled() {
+			qtL = qt.NewChild("query for tenant: %s", sq.TenantTokens[i].String())
+			defer qtL.Done()
```

Comment:
Wouldn't that potentially introduce issues in case there will be thousands of tenants?
AFAIK we had a ticket mentioning multitenant querying with 1k+ tenants case.

Reply:
It's not a hot path, where defer may introduce performance issues. Basically, each defer call will allocate an object on heap with pointer to the next defer statement in chain.